kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,617 @@
+
+namespace Kratos
+{
+///@addtogroup Kratos MPI Core
+///@{
+
+///@name Kratos Classes
+///@{
+
+template <class TPointerDataType, class TValueDataType>
+class GlobalPointerMapCommunicator;
+
+/**
+ * @brief Proxy class to update local and non-local data
+ *
+ * This class is used to update local and non-local data with given TLocalApplyFunctor and TNonLocalApplyFunctor
+ * TLocalApplyFunctor is used to update local TPointerDataType data, TNonLocalApplyFunctor is used to update
+ * non local global pointer map data which then will use TLocalApplyFunctor after communication to update value
+ * at the owner rank.
+ *
+ * Example:
+ * 1. Initial values
+ *      Node id   Rank    TEMPERATURE
+ *      1         0       10
+ *      2         1       20
+ *      3         2        5
+ *
+ * 2. Apply Update method with following Global Pointer map
+ *      Rank: 0
+ *      Map Key (using node id)         TEMPERATURE
+ *      1                               5
+ *      2                               10
+ *
+ *      Rank: 1
+ *      Map Key (using node id)         TEMPERATURE
+ *      1                               6
+ *      3                               2
+ *
+ *      Rank: 2
+ *      Empty Global Pointer map
+ *
+ * 3. Resulting values after SendAndApplyRemotely method call
+ *      Node id   Rank    TEMPERATURE
+ *      1         0       10 + 5 + 6 = 21
+ *      2         1       20 + 10 = 30
+ *      3         2       5 + 2 = 7
+ *
+ * Signature of TLocalApplyFunctor:
+ *      void(TPointerDataType& rPointerDataTypeObject, const TValueDataType& NewValue)
+ *
+ * Signature of TNonLocalApplyFunctor:
+ *      void(TValueDataType& rCurrentValueInNonLocalGlobalPointerMap, const TValueDataType& NewValue)
+ *
+ * @tparam TPointerDataType
+ * @tparam TValueDataType
+ * @tparam TLocalApplyFunctor
+ * @tparam TNonLocalApplyFunctor
+ */
+template<class TPointerDataType, class TValueDataType, class TLocalApplyFunctor, class TNonLocalApplyFunctor>
+class ApplyProxy
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    using ProxyType = ApplyProxy<TPointerDataType, TValueDataType, TLocalApplyFunctor, TNonLocalApplyFunctor>;
+
+    using TGPVector = GlobalPointersVector<TPointerDataType>;
+
+    using TGPDataMap = GlobalPointersUnorderedMap<TPointerDataType, TValueDataType>;
+
+    using TGPNonLocalDataMap = std::unordered_map<int, TGPDataMap>;
+
+    using TGPMapCommunicator = GlobalPointerMapCommunicator<TPointerDataType, TValueDataType>;
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    /**
+     * @brief Construct a new Apply Proxy object
+     *
+     * @param CurrentRank               Current rank
+     * @param rLocalApplyFunctor        Thread safe update lambda method
+     * @param rNonLocalApplyFunctor     Thread safe non-local gp map value update method
+     * @param rNonLocalDataMap          Non-local gp map
+     * @param rPointerCommunicator      Map communicator
+     */
+    ApplyProxy(
+        const int CurrentRank,
```

Comment:
you can get the current rank from the communicator

Reply:
Yes, I will :). I was following the samge signature as in the pointer_communicator. Thats the only reason :). I will change it.

Reply:
done :)